File: interfaces/todo_repository.py
# interfaces/todo_repository.py
from entities.todo import Todo
from abc import ABC, abstractmethod

class TodoRepositoryInterface(ABC):
    @abstractmethod
    def next_id(self) -> int: pass

    @abstractmethod
    def save(self, todo: Todo) -> None: pass

    @abstractmethod
    def list_all(self) -> list[Todo]: pass

File: aaa-practic/v6/interfaces/todo_repository.rb
# Ruby doesn't have abstract classes like Python's ABC
# We'll use a module to define the contract
module TodoRepository
  def next_id
    raise NotImplementedError, "#{self.class} must implement #next_id"
  end
  
  def add(todo)
    raise NotImplementedError, "#{self.class} must implement #add"
  end
  
  def list_all
    raise NotImplementedError, "#{self.class} must implement #list_all"
  end
  
  def find_by_id(todo_id)
    raise NotImplementedError, "#{self.class} must implement #find_by_id"
  end
  
  def update(todo)
    raise NotImplementedError, "#{self.class} must implement #update"
  end
  
  def delete(todo_id)
    raise NotImplementedError, "#{self.class} must implement #delete"
  end
  
  def count
    raise NotImplementedError, "#{self.class} must implement #count"
  end
end
